File: 2015/nate/day10/day10.py
import time
from itertools import groupby

def look_and_say(sequence):
    result = []
    count = 1
    current_digit = sequence[0]
    
    for i in range(1, len(sequence)):
        if sequence[i] == current_digit:
            count += 1
        else:
            result.append(str(count) + current_digit)
            count = 1
            current_digit = sequence[i]
    
    result.append(str(count) + current_digit)
    
    return ''.join(result)

def solve_look_and_say(initial_sequence, iterations):
    current = initial_sequence
    
    for i in range(iterations):
        current = look_and_say(current)
    
    return len(current)

def test_examples():
    """Test with the provided examples."""
    print("=== Testing Look-and-Say Examples ===")
    
    examples = [
        ("1", "11"),
        ("11", "21"),
        ("21", "1211"),
        ("1211", "111221"),
        ("111221", "312211")
    ]
    
    for input_seq, expected in examples:
        result = look_and_say(input_seq)
        passed = result == expected
        print(f"{input_seq} -> {result} (expected: {expected}) {'✅' if passed else '❌'}")

def solve_part1():
    """Solve Part 1 of the problem."""
    initial_sequence = "1113222113"
    iterations = 40
    
    print("\n=== Part 1: Look-and-Say Sequence ===")
    print(f"Initial sequence: {initial_sequence}")
    print(f"Iterations: {iterations}")
    
    start_time = time.perf_counter()
    final_length = solve_look_and_say(initial_sequence, iterations)
    end_time = time.perf_counter()
    
    print(f"Final sequence length: {final_length}")
    print(f"Execution time: {(end_time - start_time) * 1000:.2f}ms")
    
    return final_length

def solve_part2():
    """Solve Part 2 of the problem (50 iterations)."""
    initial_sequence = "1113222113"
    iterations = 50
    
    print("\n=== Part 2: Look-and-Say Sequence (50 iterations) ===")
    print(f"Initial sequence: {initial_sequence}")
    print(f"Iterations: {iterations}")
    
    start_time = time.perf_counter()
    final_length = solve_look_and_say(initial_sequence, iterations)
    end_time = time.perf_counter()
    
    print(f"Final sequence length: {final_length}")
    print(f"Execution time: {(end_time - start_time) * 1000:.2f}ms")
    
    return final_length

def analyze_growth():
    """Analyze how the sequence grows over iterations."""
    print("\n=== Growth Analysis ===")
    initial_sequence = "1113222113"
    
    current = initial_sequence
    for i in range(10):  # Show first 10 iterations
        length = len(current)
        print(f"Iteration {i}: length = {length}")
        current = look_and_say(current)

if __name__ == "__main__":
    test_examples()
    solve_part1()
    solve_part2()
    analyze_growth() 

File: 2015/nate/day10/day10.js
function lookAndSay(sequence) {
    let result = '';
    let count = 1;
    let currentDigit = sequence[0];
    
    for (let i = 1; i < sequence.length; i++) {
        if (sequence[i] === currentDigit) {
            count++;
        } else {
            result += count + currentDigit;
            count = 1;
            currentDigit = sequence[i];
        }
    }
    
    result += count + currentDigit;
    
    return result;
}

function solveLookAndSay(initialSequence, iterations) {
    let current = initialSequence;
    
    for (let i = 0; i < iterations; i++) {
        current = lookAndSay(current);
    }
    
    return current.length;
}

function testExamples() {
    console.log("=== Testing Look-and-Say Examples ===");
    
    const examples = [
        { input: "1", expected: "11" },
        { input: "11", expected: "21" },
        { input: "21", expected: "1211" },
        { input: "1211", expected: "111221" },
        { input: "111221", expected: "312211" }
    ];
    
    examples.forEach(({ input, expected }) => {
        const result = lookAndSay(input);
        const passed = result === expected;
        console.log(`${input} -> ${result} (expected: ${expected}) ${passed ? '✅' : '❌'}`);
    });
}

function solve(iterations) {
    const initialSequence = "1113222113";
    
    console.log("\n=== Part 1: Look-and-Say Sequence ===");
    console.log(`Initial sequence: ${initialSequence}`);
    console.log(`Iterations: ${iterations}`);
    
    const startTime = performance.now();
    const finalLength = solveLookAndSay(initialSequence, iterations);
    const endTime = performance.now();
    
    console.log(`Final sequence length: ${finalLength}`);
    console.log(`Execution time: ${(endTime - startTime).toFixed(2)}ms`);
    
    return finalLength;
}


if (require.main === module) {
    testExamples();
    solve(40);
    solve(50);
}

module.exports = { lookAndSay, solveLookAndSay }; 